### ISSUE TITLE
ctests/example can't find function calls

### ISSUE THREAD

**Issue by jlombs** [uwhpsc-2016/homework2#7]
Maybe it's just me, but I can't seem to get the example function to compile. My functions and headers are defined, and I can test everything smoothly from the Python wrapped invocations and test file, but when I try to compile and link a new c file with the shared object library that we created it claims that the function calls are "undefined references". 

I'm following the compilation command included in the comment of the example file. I've tried fiddling around, and can't seem to get it to find the vec_add definition (or anything else for that matter). It'll compile just fine if I remove any references to functions defined in our library, but that kills the purpose of it. 

Is there something else we need to include that I'm missing? Do we need a separate header for libhomework.so, or is there a special way to invoke the functions included in that library?   

Thanks in advance, @cswiercz

**Comment by jhyearsley** [comment]
I was having a similar problem running the example file. I don't think we are having the exact same issue, but what solved mine was to get rid of the semicolons after the function definitions in the linalg.h file. So previously the function vec_add read:
"void vec_add(double* out, double* v, double* w, int N);"

and I changed it to:
"void vec_add(double* out, double* v, double* w, int N)"

That solved my problem and I am now able to run the example test, but again I wasn't receiving an error about undefined references, instead it was telling me that there was an expected identifier error.

**Comment by jlombs** [comment]
Still a no-go. Everything appears to be in the right directory, and the compilation doesn't have any problem finding the header files (it complains if I move/remove them, or don't include them in the linking), but it won't find the actual function definitions (even though the python scripts find them just fine).

**Comment by rachka** [comment]
I'm having the same issue as jlombs. The readme.md file says "everything you need to write for this homework will be put in the files src/linalg.h and src/solvers.h". I feel as though that is a typo and should be src/linalg.c and src/solvers.c based on context. Also I think the header files were supposed to be in the include/ directory.

**Comment by mvelegar** [comment]
There is a typo in `README.md`, it should read: 

Aside from writing your own tests and performing computations for your report, everything you need to write for this homework will be put in the files `src/linalg.c` and `src/solvers.c`. 

Please do not modify the header files (with the `.h` extension). @cswiercz could you please make this change and send instructions for the class?

**Comment by jlombs** [comment]
@rachka While it's true that the readme is a little misleading, I don't think that has anything to do with the issue in this thread. All of my files are where they're supposed to be.

@mvelegar Aren't we supposed to add our documentation to the header files in the include directory? That's what they're for--to provide the prototypes and to give one concise location for the user to read about the usage of the library functions. So we should be modifying the .h headers--just with comments/documentation and nothing else. All of the function definitions belong in the .c functions in the src directory, of course. 

Still, I don't think that has any bearing on the issue I opened :/

**Comment by mvelegar** [comment]
@jlombs Yes you are right, you need to add documentation in the header files. I am sorry, I meant do not add your code here, add it to the files `src/linalg.c` and `src/solvers.c`.

Please try the following command to compile your example file:

`
gcc -Llib -Iinclude -lhomework2 --shared -fPIC -Wall ./ctests/example.c -o example
`

Let me know if that works!

**Comment by shinwookang** [comment]
@mvelegar for me, when I run  `gcc -Llib -Iinclude -lhomework2 --shared -fPIC -Wall ./ctests/example.c -o example`, it gives me: Segmentation fault (core dumped)

I'm jumping on this thread, because I too am having hard time even running the ctests. I actually began coding without testing anything. 

As far as I understand, vec_add example is perfectly written out for us (I didn't touch any code), as is, but like @jlombs, but the compiler may be missing something. Maybe we could go over this example in class on Tuesday?

**Comment by jlombs** [comment]
@mvelegar So I tried that command--I tried something similar myself earlier (without the -Wall) flag, and got the same results: it'll compile, but I always get a seg fault. 

I tested by commenting out any references to the library files (the include any function calls) and just sticking in a basic print function. Using the command you listed above, it'll compile but won't execute without a seg fault. If I compile it without any of the --shared --fPIC flags, it'll compile and NOT give a seg fault, all good. 

So there's something about those flags that is inducing a seg fault that has nothing to do with any of the stuff we've coded.

**Comment by mvelegar** [comment]
Hold on guys, in midst of trying to figure it out.

**Comment by mvelegar** [comment]
Hi guys,

There is a problem using the shared-object file (libhomework2.so) which I'm not able to figure out at this time ( @cswiercz please address ), for now you can link all sources statically using this:

`
gcc -Iinclude ctests/example.c src/linalg.c src/solvers.c -o example
`

Let me know if that works.

**Comment by shinwookang** [comment]
That seems to work. How has anyone else been able to test ctests?

**Comment by jlombs** [comment]
@mvelegar actually that still doesn't work for me 100%. It's certainly closer, but I'm getting an undefined reference for functions that use the <math> library. Does that need to be sourced separately, or included in the 'include' for the static linking to find it?

**Comment by mvelegar** [comment]
@jlombs you should not need to do anything extra, I was able to compile and run `example` without any changes to `example.c` or any header/source files. The declarations for the functions are loaded from the `linalg.h` and `solver.h` files, which are `#include`-ed in `example.c`, while the implementations are picked up at compile time as we listed the corresponding source files in the `gcc` command.

Have you made any changes to header/source/test files? If you have added comments (documentation) to header files, you might want to double-check that you did not accidentally comment out code. If that still does not resolve your problem, please post the exact command you ran and the full output of gcc. If you've made any changes to code, please push those changes to your private repo so I can take a look.

**Comment by jlombs** [comment]
@mvelegar Everything is already pushed to my private repo so you can take a look. I've finished the assignment already, so all of the header/source/test files are edited, and `make lib` and `make test` run everything without any errors for the tests I've written. So the unedited version would compile fine using the command you mentioned, but that's because the `math` library and function calls using it aren't in the starting repo, but that's where it's faulting for my repo.

It's weird that it's not finding the reference to the math functions when I statically link, because it finds it just fine when linking the shared object library. 

The command I'm using is the one you suggested, with one small modification: 

` gcc -std=c99 -Iinclude ctests/example.c src/linalg.c src/solvers.c -o ctests/example`

I've been compiling everything with a force flag to use c99, since all of the inline loop declarations require it, but my laptop doesn't use that by default. You'll notice I edited the compiler flag in the Makefile as well to reflect that, so everything is using the same standard. You can pop off that flag in the makefile and in the command above if you don't need it--I tried this in the SMC without those flags and get the same results, so that's not effecting anything.

Thanks for taking a look!

**Comment by cswiercz** [comment]
> Aside from writing your own tests and performing computations for your report, everything you need to write for this homework will be put in the files src/linalg.c and src/solvers.c.

This has been corrected in the `README.md`. Pull the changes so it will reflect in your own copies.

To clarify:

* `src/` - write your solutions (the function bodies) in the `.c` files provided here
* `include/` - the `.h` header files here should only contain function prototypes as well as the function documentation.

> for me, when I run gcc -Llib -Iinclude -lhomework2 --shared -fPIC -Wall ./ctests/example.c -o example, it gives me: Segmentation fault (core dumped)

Figuring this out. Will update makefile accordingly.

**Comment by cswiercz** [comment]
# Compiling `example.c` Against `libhomework2.so`

Strange. The the `example.c` (which is not necessary to use when working on this homework) compiles, links, and executes fine on my own machine but not on SMC. Here is what happens with a fresh clone of the homework2 repo on SMC. I take it some of you are experiencing similar issues:

```
~$ git clone https://github.com/uwhpsc-2016/homework2.git
Cloning into 'homework2'...
remote: Counting objects: 36, done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 36 (delta 9), reused 35 (delta 8), pack-reused 0
Unpacking objects: 100% (36/36), done.
Checking connectivity... done.

~$ cd homework2

homework2$ ls
ctests  homework2  include  lib  Makefile  README.md  Scratch.ipynb  src  test_homework2.py

homework2$ make
gcc -Iinclude -c -Wall -fPIC src/linalg.c -o src/linalg.o
gcc -Iinclude -c -Wall -fPIC src/solvers.c -o src/solvers.o
gcc --shared src/linalg.o src/solvers.o -o lib/libhomework2.so

homework2$ python test_homework2.py
 
===== Running Tests =====
test_vec_add (__main__.TestLinalg) ... ok
 
----------------------------------------------------------------------
Ran 1 test in 0.000s
 
OK

homework2$ gcc ctests/example.c -Llib -lhomework2 -Iinclude -o example

homework2$ ./example
./example: error while loading shared libraries: libhomework2.so: cannot open shared object file: No such file or directory

homework2$  
```

*Conclusion:* the Python wrappers aren't having a problem linking but gcc *is*. I will continue working this out.

**Comment by cswiercz** [comment]
## Aha! A Solution

**Short version:**

Add the flag `-Wl,-rpath,./lib` to the `example.c` compile:

```
$ gcc ctests/example.c -lhomework2 -Iinclude -Wl,-rpath,./lib -o example
```

**Comment by jlombs** [comment]
@cswiercz For me, it'll find the dynamic library now, but it won't take any of the other non-explicitly defined functions with it (same with the static link). The current error is : 
`lib/libhomework2.so: undefined reference to 'pow'`, where 'pow' is a function from the \<math> include... Tried it on SMC as well as my local machine

**Comment by cswiercz** [comment]
`pow` is defined in `<math.h>`.

**Comment by jlombs** [comment]
I know. Do I have to link that separately or include it separately? It's included in my linalg.c function, and the python wrapped invocation works fine, but gcc won't find it when I compile with the command you gave...

**Comment by cswiercz** [comment]
@jlombs See above.

**Comment by jlombs** [comment]
Repeating the code you used will compile fine. It's when I call `pow` from within a function defined within my `linalg` library that it gives an issue. I have zero clue why. It's well defined in its inputs because I use explicit numbers and not variable references. Also, like I said, it compiles just fine when I run my make test or make lib functions. Only when I try to link the library does it complain that pow isn't defined anymore

**Comment by cswiercz** [comment]
Did you include `<math.h>` at the top of `linalg.c`?

You are correct in that C will coerce literals ("explicit numbers") into the appropriate types.

Does it work when you run the Python test suite?

**Comment by jlombs** [comment]
Yes, it's included at the top. And also yes, as I said, the python test suite runs without any issue. That's why this is weird. It's somehow forgetting about that include when we go to re-link everything from the shared library

**Comment by cswiercz** [comment]
This is strange. Can you push some "non-working" code to your repo so I can take a look?

**Comment by jlombs** [comment]
I actually did just that 2 days ago with literally that commit message, haha. "Modified example test to minimal non-working example...can't get compiler to find the function definitions!"

Anyway, thanks for taking a look

**Comment by shinwookang** [comment]
just to piggyback, my issue now is, when I run `gcc ctests/example.c -lhomework2 -Iinclude -Wl,-rpath,./lib -o example` is `/usr/bin/ld: cannot find -lhomework2
collect2: error: ld returned 1 exit `status``

When I just do `gcc ctests/example.c -Llib -lhomework2 -Iinclude -o example`, my error message is `lib/libhomework2.so: undefined reference to `sqrt'
collect2: error: ld returned 1 exit status` instead of 'pow' like @jlombs.

FYI, all my `linalg.c` functions are tested within the test suite and work fine.

**Comment by jlombs** [comment]
@cswiercz may have made a typo in his Short Version solution. You still need the -Llib flag, I believe. (At least, I've been including it, or else I get the same error). 

But yeah, then we're still stuck not being able to use math functions it looks like

**Comment by cswiercz** [comment]
Darn, another typo.

I'm looking into this.

**Comment by rachka** [comment]
@cswiercz @jlombs @shinwookang I'm not sure if this would solve the issues that you are having but I was able to link the library to example.c without the undefined reference error by using the following command

```gcc ctests/example.c -Llib -lhomework2 -Iinclude -Wl,-rpath,./lib -lm -o example```

It seems that the -lm flag is needed to tell gcc to link your code against the math lib. Apparently placing it after the objects you want to link does the trick. The link at stackoverflow is 
http://stackoverflow.com/questions/8671366/undefined-reference-to-pow-and-floor

**Comment by cswiercz** [comment]
@rachka Thank you for looking this up! It must be a very machine / compiler dependent behavior. On my own machine I don't think I have ever needed to explicitly link to `libm`. Then again, I'm using `clang` instead of `gcc`.

I will modify `example.c` documentation.

**Comment by cswiercz** [comment]
I will also say something about this tomorrow morning.

**Comment by jlombs** [comment]
In case anyone else still doesn't compile after reading this (like mine), the fix for me turned out to be all of the above, PLUS when you specify the runtime library path, it needed to be specified from the executable directory, not from the compile directory. 

No `./` shortcuts. Which I guess makes sense, seeing that at runtime, the location of `./` will be wherever you put your executable, although I figured `gcc` was smart enough to take that into account at specification. 

My executable was in the ctests directory, not the main directory, and as a result, fully specifying the path or even doing `../lib` is necessary.
